cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -511,50 +536,112 @@ func (scaleSet *ScaleSet) Nodes() ([]cloudprovider.Instance, error) {
 	lastRefresh := time.Now().Add(-time.Second * time.Duration(splay))
 
+	orchestrationMode, err := scaleSet.getOrchestrationMode()
+	if err != nil {
+		klog.Errorf("failed to get information for VMSS: %s, error: %v", scaleSet.Name, err)
+		return nil, err
+	}
+
+	klog.V(4).Infof("VMSS: orchestration Mode %s", orchestrationMode)
+
+	if orchestrationMode == compute.Uniform {
+		err := scaleSet.buildScaleSetCache(lastRefresh)
+		if err != nil {
+			return nil, err
+		}
+
+	} else if orchestrationMode == compute.Flexible {
+		if scaleSet.manager.config.EnableVmssFlex {
+			err := scaleSet.buildScaleSetCacheForFlex(lastRefresh)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("vmss - %q with Flexible orchestration detected but 'enableVmssFlex' feature flag is turned off", scaleSet.Name)
+		}
+
+	} else {
+		return nil, fmt.Errorf("Failed to determine orchestration mode for vmss %q", scaleSet.Name)
+	}
+
+	klog.V(4).Infof("Nodes: returns")
+	return scaleSet.instanceCache, nil
+}
+
+func (scaleSet *ScaleSet) buildScaleSetCache(lastRefresh time.Time) error {
 	vms, rerr := scaleSet.GetScaleSetVms()
 	if rerr != nil {
 		if isAzureRequestsThrottled(rerr) {
 			// Log a warning and update the instance refresh time so that it would retry after cache expiration
 			klog.Warningf("GetScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
 			scaleSet.lastInstanceRefresh = lastRefresh
-			return scaleSet.instanceCache, nil
+			return nil
 		}
-		return nil, rerr.Error()
+		return rerr.Error()
 	}
 
 	scaleSet.instanceCache = buildInstanceCache(vms)
 	scaleSet.lastInstanceRefresh = lastRefresh
-	klog.V(4).Infof("Nodes: returns")
-	return scaleSet.instanceCache, nil
+
+	return nil
+}
+
+func (scaleSet *ScaleSet) buildScaleSetCacheForFlex(lastRefresh time.Time) error {
+	vms, rerr := scaleSet.GetFlexibleScaleSetVms()
+	if rerr != nil {
+		if isAzureRequestsThrottled(rerr) {
+			// Log a warning and update the instance refresh time so that it would retry after cache expiration
+			klog.Warningf("GetFlexibleScaleSetVms() is throttled with message %v, would return the cached instances", rerr)
+			scaleSet.lastInstanceRefresh = lastRefresh
+			return nil
+		}
+		return rerr.Error()
+	}
+
+	scaleSet.instanceCache = buildInstanceCache(vms)
+	scaleSet.lastInstanceRefresh = lastRefresh
+
+	return nil
 }
 
 // Note that the GetScaleSetVms() results is not used directly because for the List endpoint,
 // their resource ID format is not consistent with Get endpoint
-func buildInstanceCache(vms []compute.VirtualMachineScaleSetVM) []cloudprovider.Instance {
+func buildInstanceCache(vmList interface{}) []cloudprovider.Instance {
 	instances := []cloudprovider.Instance{}
 
-	for _, vm := range vms {
-		// The resource ID is empty string, which indicates the instance may be in deleting state.
-		if len(*vm.ID) == 0 {
-			continue
+	switch vms := vmList.(type) {
+	case []compute.VirtualMachineScaleSetVM:
```

Comment:
Having to switch on type here looks a little weird, given that much else follows separate codepaths for Flex vs. Uniform. I think I would have preferred either more push for shared code elsewhere, or full separation here as well. But not a big deal, does not have to be addressed.